### PR TITLE
[MAINTENANCE] Start adding shared fields to Expectation model

### PR DESCRIPTION
**Pull request by cdkini**
Start adding fields to Expectation models

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/expectations/expectation.py
Comment (cdkini) [inline great_expectations/expectations/expectation.py]: Not so sure about these just yet

File: great_expectations/core/result_format.py
Comment (cdkini) [inline great_expectations/core/result_format.py]: Don't fully know the shape of this but this will provide some basic structure until we refactor/rework result format

File: great_expectations/expectations/core/expect_column_values_to_be_increasing.py
Comment (cdkini) [inline great_expectations/expectations/core/expect_column_values_to_be_increasing.py]: This is done quite often in our codebase - no need to override if you're just going to leverage the parent impl

File: great_expectations/expectations/expectation.py
Comment (cdkini) [inline great_expectations/expectations/expectation.py:1261]: Would fully delete this but until we 100% shift over to Pydantic validation, we'll require this method on the base class to perform our hand-rolled validation. Child classes still depend on the existence of this method in the hierarchy.

File: great_expectations/expectations/expectation.py
Comment (cdkini) [inline great_expectations/expectations/expectation.py:2537]: Deleted the deprecated classes while I was in here

File: great_expectations/expectations/expectation.py
Comment (cdkini) [inline great_expectations/expectations/expectation.py:2702]: Starting to chip away at remove this hand-rolled validation funcs within the base classes

File: great_expectations/expectations/expectation.py
Comment (cdkini) [inline great_expectations/expectations/expectation.py]: Both these fields are irrelevant moving forward - I'm going to delete them as I go

File: great_expectations/warnings.py
Comment (cdkini) [inline great_expectations/warnings.py]: Note to self - raise error in this method and delete whatever fails

File: tests/expectations/core/test_expect_column_max_to_be_between.py
Comment (cdkini) [inline tests/expectations/core/test_expect_column_max_to_be_between.py:12]: All these tests test for a path we no longer support - just deleting them outright

File: tests/expectations/test_expectation.py
Comment (cdkini) [inline tests/expectations/test_expectation.py:172]: A breaking change but we should leverage Pydantic's type instead of our own configuration error if we're going to be using models

File: tests/expectations/test_generate_diagnostic_checklist.py
Comment (cdkini) [inline tests/expectations/test_generate_diagnostic_checklist.py:34]: Feels iffy but all these diagnostic methods are instance methods - I have a ticket to try and make them classmethods so don't need to pass meaningless params but that's less of a priority atm

File: great_expectations/expectations/expectation.py
Comment (cdkini) [inline great_expectations/expectations/expectation.py]: Originally just used `Field` with `ge` and `le` but that didn't seem to mesh well with inheritance?

This validator works with child classes but I saw test failures with the other pattern - worth investigating further.

File: great_expectations/expectations/expectation.py
Comment (cdkini) [inline great_expectations/expectations/expectation.py]: Weirdly enough, `pydantic.Field` and the in-built validation done through `gt` and `lt` didn't mesh well with inheritance

Comment (Kilo59) [reply great_expectations/expectations/expectation.py]: Have you tried using pydantic constrained field types? 
https://docs.pydantic.dev/1.10/usage/types/#constrained-types

File: great_expectations/expectations/expectation.py
Comment (Kilo59) [reply great_expectations/expectations/expectation.py]: Validators don't run by default for fields that aren't present.
https://docs.pydantic.dev/1.10/usage/validators/#validate-always
There are other scenarios which can cause vaildators not to run on subclasses. But that doesn't look like it applies in this case.
https://docs.pydantic.dev/1.10/usage/validators/#subclass-validators-and-each_item

------

IMO the best thing todo would be to not rely on validators and use a `confloat` or custom type.
https://docs.pydantic.dev/1.10/usage/types/#arguments-to-confloat
Using the `confloat` has the extra benefit of encoding the valid results to any generated jsonschema.

```python
from typing import Final, Type
from pydantic.v1 import BaseModel, confloat

ZeroToOne: Final[Type] = confloat(ge=0, le=1)

class Foo(BaseModel):
  foo_field: ZeroToOne

class Bar(BaseModel):
  bar_field: ZeroToOne
```

File: tests/expectations/test_expectation.py
Comment (Kilo59) [reply tests/expectations/test_expectation.py:172]: `non-blocking`
pydantic validation errors happen on `__init__`. So we could have a custom `Expectation.__init__` that catches and rethrows `pydantic.ValidationError`s as `InvalidExpectationConfigurationError`.

But I also don't think we should be too concerned with breaking changes for the V1 release.

File: great_expectations/validator/validator.py
Comment (Kilo59) [inline great_expectations/validator/validator.py:1125]: Is `evaluated_config: ExpectationConfiguration` a pydantic model now as well?

**Review by Kilo59 — APPROVED** [review]
I don't know the Expectation code that well so I would get a second opinion but LGTM.

File: great_expectations/validator/validator.py
Comment (cdkini) [reply great_expectations/validator/validator.py:1125]: Nope we're going to get rid of that entirely

File: tests/expectations/test_expectation.py
Comment (cdkini) [reply tests/expectations/test_expectation.py:172]: Yeah I think we're okay with `pydantic.ValidationError` and completely moving away from our custom error - good call though!

**Pull request merged**
